mime.go

```diff
@@ -116,107 +122,193 @@ func (mh *mimeHandler) initJPEGEncoder(frame *C.AVFrame) error {
 		}
 	}
-	if mh.yuyvSwsCtx == nil {
-		return nil, camera.ImageMetadata{}, errors.New("failed to create YUYV converter")
+
+	if *swsCtx == nil {
+		return nil, camera.ImageMetadata{}, errors.New("failed to create " + format + " converter")
 	}
-	if mh.yuyvFrame == nil {
-		return nil, camera.ImageMetadata{}, errors.New("failed to create yuyv destination frame")
+	if *dstFrame == nil {
+		return nil, camera.ImageMetadata{}, errors.New("failed to create " + format + " destination frame")
 	}
+
 	res := C.sws_scale(
-		mh.yuyvSwsCtx,
+		*swsCtx,
 		(**C.uint8_t)(unsafe.Pointer(&frame.data[0])),
 		(*C.int)(unsafe.Pointer(&frame.linesize[0])),
 		0,
 		frame.height,
-		(**C.uint8_t)(unsafe.Pointer(&mh.yuyvFrame.data[0])),
-		(*C.int)(unsafe.Pointer(&mh.yuyvFrame.linesize[0])),
+		(**C.uint8_t)(unsafe.Pointer(&(*dstFrame).data[0])),
+		(*C.int)(unsafe.Pointer(&(*dstFrame).linesize[0])),
 	)
 	if res < 0 {
-		return nil, camera.ImageMetadata{}, newAvError(res, "failed to convert frame to YUYV")
-	}
-	yuyvDataSize := int(mh.yuyvFrame.width) * int(mh.yuyvFrame.height) * yuyvBytesPerPixel
-	header := packYUYVHeader(int(mh.yuyvFrame.width), int(mh.yuyvFrame.height))
-	// Preallocate the final slice with the combined size of the header and the image data
-	yuyvPacket := make([]byte, len(header)+yuyvDataSize)
-	copy(yuyvPacket[0:], header)
-	// Copy the YUYV data directly into the preallocated slice
-	C.memcpy(unsafe.Pointer(&yuyvPacket[len(header)]), unsafe.Pointer(mh.yuyvFrame.data[0]), C.size_t(yuyvDataSize))
-
-	return yuyvPacket, camera.ImageMetadata{
-		MimeType: mimeTypeYUYV,
+		return nil, camera.ImageMetadata{}, newAvError(res, "failed to convert frame to "+format)
+	}
+
+	dataSize := int((*dstFrame).width) * int((*dstFrame).height) * bytesPerPixel
+	header := packHeader(format, int((*dstFrame).width), int((*dstFrame).height), headerDimBytes)
+	data := make([]byte, len(header)+dataSize)
+	copy(data[0:], header)
+	C.memcpy(unsafe.Pointer(&data[len(header)]), unsafe.Pointer((*dstFrame).data[0]), C.size_t(dataSize))
+
+	return data, camera.ImageMetadata{
+		MimeType: mimeType,
 	}, nil
 }
 
-func (mh *mimeHandler) initYUYVContext(frame *C.AVFrame) error {
-	mh.logger.Info("creating YUYV sws context with frame size: ", frame.width, "x", frame.height)
-	if mh.yuyvSwsCtx != nil {
-		C.sws_freeContext(mh.yuyvSwsCtx)
+func (mh *mimeHandler) convertYUYV(frame *C.AVFrame) ([]byte, camera.ImageMetadata, error) {
+	return mh.convertPixelFormat(
+		frame,
+		yuyvMagicString,
+		&mh.yuyvSwsCtx,
+		&mh.yuyvFrame,
+		&mh.yuyvMu,
+		mh.initYUYVContext,
+		yuyvBytesPerPixel,
+		yuyvHeaderDimBytes,
+		mimeTypeYUYV,
+	)
+}
+
+func (mh *mimeHandler) convertRGBA(frame *C.AVFrame) ([]byte, camera.ImageMetadata, error) {
+	return mh.convertPixelFormat(
+		frame,
+		rgbaMagicString,
+		&mh.rgbaSwsCtx,
+		&mh.rgbaFrame,
+		&mh.rgbaMu,
+		mh.initRGBAContext,
+		rgbaBytesPerPixel,
+		rgbaHeaderDimBytes,
+		rutils.MimeTypeRawRGBA,
+	)
+}
```

Comment:
can we re-organize this code so that the YUV and RGBA handlers are close to were they're first used, plus grouped together where appropriate.